### ISSUE TITLE
💅 noMisusedPromises should emit warning where function returns `boolean | Promise<boolean>`

### ISSUE THREAD

**Issue by himself65** [biomejs/biome#8405]
### Environment information

```bash
CLI:
  Version:                      2.3.8
  Color support:                true

Platform:
  CPU Architecture:             aarch64
  OS:                           macos

Environment:
  BIOME_LOG_PATH:               unset
  BIOME_LOG_PREFIX_NAME:        unset
  BIOME_CONFIG_PATH:            unset
  BIOME_THREADS:                unset
  NO_COLOR:                     unset
  TERM:                         xterm-256color
  JS_RUNTIME_VERSION:           v22.18.0
  JS_RUNTIME_NAME:              node
  NODE_PACKAGE_MANAGER:         unset

Biome Configuration:
  Status:                       Loaded successfully
  Path:                         biome.json
  Formatter enabled:            true
  Linter enabled:               true
  Assist enabled:               true
  VCS enabled:                  false

Linter:
  JavaScript enabled:           unset
  JSON enabled:                 unset
  CSS enabled:                  unset
  GraphQL enabled:              unset
  Recommended:                  false
  Enabled rules:
    complexity/noUselessSwitchCase
    complexity/noUselessTypeConstraint
    complexity/useDateNow
    correctness/noUnusedImports
    nursery/noFloatingPromises
    nursery/noMisusedPromises
    performance/noDelete
    style/noRestrictedImports
    style/useImportType
    style/useNodejsImportProtocol
    suspicious/noDebugger
    suspicious/noDuplicateObjectKeys
    suspicious/noImplicitAnyLet
    suspicious/noTsIgnore

Workspace:
  Open Documents:               0
```

### Rule name

noMisusedPromises

### Playground link

https://biomejs.dev/playground/?lintRules=noMisusedPromises&code=ZgB1AG4AYwB0AGkAbwBuACAAYQAoACkAOgAgAGIAbwBvAGwAZQBhAG4AIAB8ACAAUAByAG8AbQBpAHMAZQA8AGIAbwBvAGwAZQBhAG4APgAgAHsAfQAKAAoAaQBmACAAKABhACgAKQApACAAewAKACAAIAAKAH0ACgAKAGYAdQBuAGMAdABpAG8AbgAgAGIAKAApADoAIABQAHIAbwBtAGkAcwBlADwAYgBvAG8AbABlAGEAbgA%2BACAAewB9AAoACgBpAGYAIAAoAGIAKAApACkAIAB7AAoAIAAgAAoAfQA%3D

### Expected result

should emit an error or a warning

### Code of Conduct

- [x] I agree to follow Biome's Code of Conduct

**Comment by himself65** [comment]
Related: https://github.com/better-auth/better-auth/pull/6611

**Comment by taga3s** [comment]
I'd like to work on this issue.
Could you assign it to me?

**Comment by siketyan** [comment]
@taga3s all yours!
